src/hist/named.py

```diff
@@ -12,6 +13,25 @@ def __init__(self, *args, **kwargs):
         if "" in self.names:
             raise Exception("Each axes in the NamedHist instance should have a name.")
 
+    def project(self, *args: Tuple[str]):
+        """
+        Projection of axis idx.
+        """
+
+        indices: tuple = tuple()
+        for name in args:
+            find = False
+            for index, axis in enumerate(self.axes):
+                if name == axis.name:
+                    indices += (index,)
+                    find = True
+                    break
+
+            if not find:
+                raise ValueError("The axis names could not be found.")
+
+        return super(Histogram, self).project(*indices)
```

Comment:
**How to skip the `BaseHist` and directly use `bh.Histogram`**?

Reply:
When using `super().project(*indices)`, pytest fails.

Reply:
Super goes one up. If you want to go up further, just call it directly. So `Histogram.project(self, *indices)`. However, you don't actually want that; NamedHist should: check to see if all arguments are strings - if one or more are ints, it should fail. Then it should just pass on to Hist: `super().project(*args)`. Hist should take strings or ints.

Reply:
Yeah, we should allow NamedHist to use indices to project, which is different from the fill method. This is because we need to use super(), but NamedHist must be derived from BaseHist. Seems like Python's derivation must be an addition, and NamedHist has to accept the feature of its father class if we want to succeed the features of its grandfather.

Reply:
Ah, at last. Find the problem. it's risky to do that.

> NamedHist should check to see if all arguments are strings - if one or more are ints, it should fail.

**We must allow projection by the index of NamedHist**, if we don't plan to change anything of underlying BH or get_item. That's because get_item will use this property. More generally speaking, Python's derivation is an addition, if we are forced to make some minus, then some manners might be broken. 

So the current idea is, BasedHist allows projection by the index, and NamedHist **add** the feature that allows projection by the names.

Reply:
NamedHist *only* forces names.  Normal Hist prefers names, but works with both.

Reply:
> That's because get_item will use this property.

No, it calls project on the c++ class (unless it doesn't, in which case it's a bug), so it's okay if the python class doesn't support integers. But that is why you have to force this on MyPy; it is unexpected if a child class doesn't accept something a parent does. But in this case, I think it's okay.

Reply:
https://github.com/scikit-hep/boost-histogram/blob/a12244f5daf81d5e619cefdd70d0d2ad859ca39d/src/boost_histogram/_internal/hist.py#L574

This line returns the c++ class, and not a python wrapper.

Reply:
Static types would have made this easier to read in boost histogram!

Reply:
I made a new commit https://github.com/scikit-hep/hist/pull/22/commits/4661c9a840bc0df84807010672850f1e05364990 then and you can see that we cannot make subtraction for `NamedHist.project()`.

Reply:
This road seems to be blocked.

Reply:
Bug in boost-histogram then. Let's disable the check for no ints for one (in NamedHist) and tomorrow I'll fix the problem in boost-histogram. Add a comment with a TODO in the Hist src.

Reply:
Okay!

Reply:
Don't you think disable projection by names for BaseHist is more reasonable? As methods related to names seem to be exclusive for NamedHist.

Reply:
Looks like I wrapped it here: https://github.com/scikit-hep/boost-histogram/blob/a12244f5daf81d5e619cefdd70d0d2ad859ca39d/src/boost_histogram/_internal/hist.py#L304. Need to think about how to do this consistently.

Reply:
Afterall, go ahead, these are all based on correct BH :)

Reply:
No! Names are a feature for all of Hist. We don't want to discourage name use. The only reason we have NamedHist is we have one user who wants to enforce names always. Most users will mix the two. Add names, but also use position when it's more convenient, like in indexing. 

Maybe we should rename NamedHist to ForcedNamedHist. :)

Reply:
Okay, I see.